packages/expo-router/src/fork/getStateFromPath.ts

```diff
@@ -206,54 +209,33 @@ function sortConfigs(a: RouteConfig, b: RouteConfig, previousSegments: string[]
     return b.routeNames.join('>').localeCompare(a.routeNames.join('>'));
   }
 
-  // If one of the patterns starts with the other, it's more exhaustive
-  // So move it up
-  if (
-    a.pattern.startsWith(b.pattern) &&
-    // NOTE(EvanBacon): This is a hack to make sure that `*` is always at the end
-    b.screen !== 'index'
-  ) {
+  /*
+   * If one of the patterns starts with the other, it is earlier in the config sorting.
+   * However, configs are a mix of route configs and layout configs
+   * e.g There will be a config for `/(group)`, but maybe there isn't a `/(group)/index.tsx`
+   *
+   * This is because you can navigate to a directory and its navigator will determine the route
+   * These routes should be later in the config sorting, as their patterns are very open
+   * and will prevent routes from being matched
+   *
+   * Therefore before we compare segment parts, we force these layout configs later in the sorting
+   *
+   * NOTE(marklawlor): Is this a feature we want? I'm unsure if this is a gimmick or a feature.
+   */
+  if (a.pattern.startsWith(b.pattern) && !b.isIndex) {
     return -1;
   }
 
-  if (b.pattern.startsWith(a.pattern) && a.screen !== 'index') {
+  if (b.pattern.startsWith(a.pattern) && !a.isIndex) {
     return 1;
```

Comment:
This two guards are the actual cause of the bug in this issue. They are only checking for `screen !== 'index'`. I think it reads better now as you don't have the variable declarations in the middle of the guards.